code/modules/library/library_admin.dm

```diff
@@ -0,0 +1,207 @@
+	var/list/data = list()
+	data["pagestate"] = page_state
+	data["booklist"] = cached_books
+	data["ckey"] = selected_ckey ? selected_ckey : "ERROR"
+	data["reports"] = reports
+
+	data["modal"] = ui_modal_data(src)
+	return data
+
+/datum/ui_module/library_manager/ui_act(action, params, datum/tgui/ui)
+	if(..())
+		return
+
+	if(ui_act_modal(action, params))
+		return
+
+	switch(action)
+		if("view_reported_books")
+			reports = list()
+			for(var/datum/cachedbook/CB in GLOB.library_catalog.get_flagged_books())
+				for(var/datum/flagged_book/report as anything in CB.reports)
+					if(!report)
+						return
+					var/datum/library_category/report_category =  GLOB.library_catalog.get_report_category_by_id(report.category_id)
+					var/report_info = list(
+						"reporter_ckey" = report.reporter,
+						"uploader_ckey" = CB.ckey,
+						"id" = CB.id,
+						"title" = CB.title,
+						"author" = CB.author,
+						"report_description" = report_category.description,
+					)
+					reports += list(report_info)
+			page_state = LIBRARY_MENU_REPORTS
+		if("delete_book")
+			if(text2num(params["bookid"])) //make sure this is actually a number
+				if(GLOB.library_catalog.remove_book_by_id(text2num(params["bookid"])))
+					log_and_message_admins("has deleted book [params["bookid"]].")
+		if("view_book")
+			if(params["bookid"])
+				view_book_by_id(text2num(params["bookid"]), ui.user)
+		if("unflag_book")
+			if(params["bookid"])
+				if(GLOB.library_catalog.unflag_book_by_id(text2num(params["bookid"])))
+					log_and_message_admins("has unflagged book [params["bookid"]].")
+		if("return")
+			page_state = LIBRARY_MENU_MAIN
+
+
+/datum/ui_module/library_manager/proc/ui_act_modal(action, list/params)
+	. = TRUE
+	var/id = params["id"] // The modal's ID
+	var/list/arguments = istext(params["arguments"]) ? json_decode(params["arguments"]) : params["arguments"]
+	switch(ui_modal_act(src, action, params))
+		if(UI_MODAL_OPEN)
+			switch(id)
+				if("specify_ssid_delete")
+					ui_modal_input(src, id, "Please input a book SSID:", null, arguments)
+				if("specify_ckey_search")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				if("specify_ckey_delete")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				else
+					return FALSE
+		if(UI_MODAL_ANSWER)
+			var/answer = params["answer"]
+			switch(id)
+				if("specify_ssid_delete")
+					if(!answer || !text2num(answer))
+						return
+					var/confirm = alert("You are about to delete book [text2num(answer)]", "Confirm Deletion", "Yes", "No")
+					if(confirm != "Yes")
+						return //we don't need to sanitize b/c removeBookyByID uses id=:id instead of like statemetns
+					if(GLOB.library_catalog.remove_book_by_id(text2num(answer)))
+						log_and_message_admins("has deleted the book [text2num(answer)].")
+				if("specify_ckey_search")
+					if(!answer)
+						return
+					var/datum/library_user_data/search_terms = new()
+					search_terms.search_ckey = answer
```

Comment:
Might want to sanitize this one as well since you'll be using it to query

Reply:
done